### ISSUE TITLE
MQTT crashes, ignoring any incoming messages.

### ISSUE THREAD

**Issue by fjpanag** [thingsboard/thingsboard#2017]
Lately I am strangling with an issue regarding the MQTT interface of Thingsboard.

I have a device that sends telemetry twice per minute, and some attributes a few times per day.

When I start up the device and the server, everything works OK. After about a day Thingsboard does not accept any data. Checking the device telemetry, no new values arrive. MQTT communication stops completely.

I believe that this is a problem with Thingsboard. For some reason MQTT crashes and discards any publications made by my device. (Of course I have triple checked that the device is alive and that it actually sends data).

Here is what I have found so far:

- Although no new data arrive and the device *appears* dead, server attribute `active` is `true`.
- Attribute `lastActivityTime` contains a valid timestamp, indicating that there is active communication with the device.
- Due to `active` attribute, no device disconnect alarm is triggered, and it is not possible to diagnose the issue automatically.
- Restarting the device, or causing a reconnect will not fix it. Communication is dead.
- **Restarting Thingsboard is the only way to recover from this situation**.  Even without touching anything on the device.
- Enabling debug on the rule engine, even on the root node, does not produce any result. It seems like messages never arrive at the rule engine. The fault is somewhere before that.
- The log file does not contain anything about this. Neither any exceptions, nor anything related to the messages sent by the devices. I keep the log open, the device sends data, Thingsboard updates `lastActivityTime` without accepting the actual telemetry data, and absolutely nothing is written to the log.
- I believe (but I haven't proved it yet) that sending a two-way RPCs towards the device, results in the command being sent, the device to respond, and the response to be discarded due to this issue. So possibly this is related to incoming messages only.

**Comment by fjpanag** [comment]
This problem originally appeared in 2.3.1. A few days ago I updated to 2.4.1.

I can now say that the problem is much worse. Initially I had this crash once in every few days. Now with 2.4.1, it happens after a few hours, or even minutes!

I performed some more tests.
- When the problem appears, Thingsboard neither sends messages, nor receives. MQTT communication stops completely.
- When the problem appears, it affects all devices connected to Thingsboard. They simultaneously stop responding, till Thingsboard server restart.
- The devices work correctly during the issue, with proper internet connection. For example NTP is working OK during the problem. Similarly connection to this specific server is also OK, as FTP connections work correctly (FTP server is hosted on the same machine).

**Comment by morpheus000** [comment]
Did you find a solution for you problem? If got something similar but my server stops accepting messages on ca 23.00 and starts listening again at 0.00, without restart of services.
In 2.4.0 there was a exception in the logs in 2.4.1 there's no exception but the same error indication.

**Comment by lostsheephelpless** [comment]
@fjpanag Hi, I have the same problem, I install Thingsboard by rpm in the CentOS ,I try to restart Thingsboard server ,it works for a while ,then it comes the problem again,do you have any good idea?

**Comment by fjpanag** [comment]
As it turned out, the script running on the first node of my rules was failing. After 3 fails it was blacklisted, and all communications towards Thingsboard were halted.

I am pretty sure that I had enabled debug logs on this node on v2.3.1, but for some reason there was nothing. On v2.4.1 I checked again, and the blacklisted script was reported.

The solution was to increase the JS sandbox time-out from the fairly low setting of 100ms up to 10secs. This solved my issue.

The blacklist policy surely needs revising IMO, as for the moment it seems to create more problems than it solves.
